source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -0,0 +1,504 @@
+    }
+}
+
+// Emphasis Match's syntax is complete when it's Captured
+// Its syntax is always correct, therefore it attemps to
+// capture until it can't capture anymore.
+// it moves two states, emphasis state and text state,
+// at each transition of state, one token is captured
+void EmphasisParser::Match(std::stringstream &stream)
+{ 
+    while (m_current_state != EmphasisState::Captured)
+    {
+        m_current_state = m_stateMachine[m_current_state](*this, stream, m_current_token);
+    }
+}
+
+/// captures text untill it see emphasis character. When it does, switch to Emphasis state
+EmphasisParser::EmphasisState EmphasisParser::MatchText(EmphasisParser &parser, std::stringstream &stream, std::string& token)
+{
+    /// MarkDown keywords
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.peek() == '-' || stream.eof())
+    {
+        parser.Flush(token);
+
+        return EmphasisState::Captured;
+    }
+
+    if (parser.IsMarkDownDelimiter(stream.peek()))
+    {
+        // encounterred first emphasis delimiter
+        parser.CaptureCurrentCollectedStringAsRegularToken();
+        DelimiterType emphasisType = EmphasisParser::GetDelimiterTypeForCharAtCurrentPosition(stream.peek()); 
+        parser.UpdateCurrentEmphasisRunState(emphasisType);
+        token += stream.get();
+        return Emphasis;
+    }
+    else
+    {    
+        parser.UpdateLookBehind(stream.peek());
+        token += stream.get();
+        return Text;
+    }
+}
+
+/// captures text untill it see none-emphasis character. When it does, switch to text state
+EmphasisParser::EmphasisState EmphasisParser::MatchEmphasis(EmphasisParser &parser, std::stringstream &stream, std::string& token)
+{
+    // key word is encountered, flush what is being processed, and have those keyword
+    // handled by ParseBlock()
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.peek() == '-' || stream.eof())
+    {
+        parser.Flush(token);
+        return EmphasisState::Captured;
+    }
+
+    /// if another emphasis delimiter is encounterred, it is delimiter run
+    if (parser.IsMarkDownDelimiter(stream.peek()))
+    {
+        DelimiterType emphasisType = EmphasisParser::GetDelimiterTypeForCharAtCurrentPosition(stream.peek());
+        if (parser.IsEmphasisDelimiterRun(emphasisType))
+        {
+            parser.UpdateCurrentEmphasisRunState(emphasisType);
+        }
+        token += stream.get();
+    }
+    /// delimiter run is ended, capture the current accumulated token as emphasis
+    else
+    {
+        parser.CaptureEmphasisToken(stream.peek(), token);
+
+        if (stream.peek() == '\\')
+        {
+            // skips escape char
+            stream.get();
+        }
+
+        parser.ResetCurrentEmphasisState();
+        parser.UpdateLookBehind(stream.peek());
+        token += stream.get();
+
+        return Text;
```

Comment:
>Text [](start = 15, length = 4)

Can you update other usages of enum values as well?

Reply:
sure


---
In reply to: [157634340](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157634340) [](ancestors = 157634340)